### PR TITLE
Enable FEATURE_MULTICASTSTUB_AS_IL for Windows x86

### LINKED ISSUE #103958: Consider to enable FEATURE_MULTICASTSTUB_AS_IL for Windows x86
Since the initial publish of coreclr, FEATURE_MULTICASTSTUB_AS_IL has been enabled on all platforms except Windows x86. Similar to #103533, the IL method stubs are easier to maintain.

Currently the IL version is functional for Windows x86, but performs worse:

```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.22631.3737/23H2/2023Update/SunValley3)
13th Gen Intel Core i9-13900K, 1 CPU, 32 logical and 24 physical cores
.NET SDK 9.0.100-preview.5.24307.3
  [Host]     : .NET 8.0.5 (8.0.524.21615), X86 RyuJIT AVX2
  Job-JGPGRX : .NET 9.0.0 (42.42.42.42424), X86 RyuJIT AVX2
  Job-CSENVN : .NET 9.0.0 (42.42.42.42424), X86 RyuJIT AVX2

Affinity=00000000000000001111111111111111
```

| Method         | Job        | Toolchain             | Mean      | Error     | StdDev    | Ratio | RatioSD |
|--------------- |----------- |---------------------- |----------:|----------:|----------:|------:|--------:|
| SingleArg      | Job-JGPGRX | \x86-main\corerun.exe |  3.744 ns | 0.0400 ns | 0.0374 ns |  1.00 |    0.00 |
| SingleArg      | Job-CSENVN | \x86-pr\corerun.exe   |  5.239 ns | 0.0316 ns | 0.0280 ns |  1.40 |    0.02 |
|                |            |                       |           |           |           |       |         |
| ManyCast       | Job-JGPGRX | \x86-main\corerun.exe | 10.001 ns | 0.2107 ns | 0.2070 ns |  1.00 |    0.00 |
| ManyCast       | Job-CSENVN | \x86-pr\corerun.exe   | 14.281 ns | 0.0928 ns | 0.0868 ns |  1.43 |    0.03 |
|                |            |                       |           |           |           |       |         |
| ManyArg_RetFPU | Job-JGPGRX | \x86-main\corerun.exe |  5.971 ns | 0.0567 ns | 0.0530 ns |  1.00 |    0.00 |
| ManyArg_RetFPU | Job-CSENVN | \x86-pr\corerun.exe   |  7.341 ns | 0.1287 ns | 0.1204 ns |  1.23 |    0.02 |

Consider to improve the codegen in JIT for better performance.

### PR DESCRIPTION
**Pull request by huoyaoyuan**
Closes #103958 .

Local benchmark result for `Action<int>` on x86:

| Method         | Job        | Toolchain                               | Mean      | Error     | StdDev    | Ratio |
|--------------- |----------- |---------------------------------------- |----------:|----------:|----------:|------:|
| SingleArg      | Job-OSFVND | \x86-3-TuneBranchPrediction\corerun.exe |  3.464 ns | 0.0127 ns | 0.0119 ns |  0.95 |
| SingleArg      | Job-KBQPJV | \x86-main\corerun.exe                   |  3.661 ns | 0.0222 ns | 0.0208 ns |  1.00 |
|                |            |                                         |           |           |           |       |
| ManyCast       | Job-OSFVND | \x86-3-TuneBranchPrediction\corerun.exe |  8.188 ns | 0.0944 ns | 0.0883 ns |  0.97 |
| ManyCast       | Job-KBQPJV | \x86-main\corerun.exe                   |  8.483 ns | 0.0888 ns | 0.0831 ns |  1.00 |
|                |            |                                         |           |           |           |       |
| ManyArg_RetFPU | Job-OSFVND | \x86-3-TuneBranchPrediction\corerun.exe |  5.566 ns | 0.0338 ns | 0.0316 ns |  0.96 |
| ManyArg_RetFPU | Job-KBQPJV | \x86-main\corerun.exe                   |  5.771 ns | 0.0212 ns | 0.0177 ns |  1.00 |

### CODE REVIEW

**Comment by huoyaoyuan** [comment]
@EgorBot -intel -amd -arm64

```csharp
using BenchmarkDotNet.Attributes;
using System;

namespace BenchmarkGround
{
    public struct GCStruct
    {
        public object a, b, c, d, e, f, h, g;
    }

    public class Bench
    {
        private readonly object obj = new object();

        private readonly Guid guid = Guid.NewGuid();

        private readonly GCStruct gcStruct = new GCStruct
        {
            a = new object(),
            g = new object()
        };

        private readonly Action<int> singleArg = (Action<int>)delegate { } + delegate { };

        private readonly Action<int> manyCast = (Action<int>)delegate { }
        +
            delegate { }
        + delegate { }
        + delegate { }
        + delegate { }
        + delegate { };

        private readonly Func<int, double, Guid, GCStruct, object, double> manyArg_RetFPU = (Func<int, double, Guid, GCStruct, object, double>)
            delegate { return 123.456; }
        + delegate { return 654.321; };

        [Benchmark]
        public void SingleArg() => singleArg(42);

        [Benchmark]
        public void ManyCast() => manyCast(42);

        [Benchmark]
        public double ManyArg_RetFPU() => manyArg_RetFPU(42, 123.0, guid, gcStruct, obj);
    }
}
```

File: src/coreclr/vm/comdelegate.cpp [hunk 1]
```diff
@@ -2189,14 +2188,37 @@ FCIMPL1(PCODE, COMDelegate::GetMulticastInvoke, Object* refThisIn)
...
+        pCode->EmitLDC(DBCF_ATTACHED);
+        pCode->EmitAND();
+        pCode->EmitBRTRUE(invokeTraceHelper);
+        pCode->EmitBR(debuggerCheckEnd); // Tune branch prediction to prefer non-debugging path
+
+        pCode->EmitLabel(invokeTraceHelper);
+
         pCode->EmitLoadThis();
         pCode->EmitLDLOC(dwLoopCounterNum);
         pCode->EmitCALL(METHOD__STUBHELPERS__MULTICAST_DEBUGGER_TRACE_HELPER, 2, 0);
```
Comment (jkotas) [inline src/coreclr/vm/comdelegate.cpp:2209]:
```suggestion
        pCode->EmitBRTRUE(invokeTraceHelper);
```

And move the debugging path to be after `RET`:

```
        pCode->EmitRET();

#ifdef DEBUGGING_SUPPORTED
        // Emit debugging support at the end of the method for better perf
        pCode->EmitLabel(invokeTraceHelper);

        pCode->EmitLoadThis();
        pCode->EmitLDLOC(dwLoopCounterNum);
        pCode->EmitCALL(METHOD__STUBHELPERS__MULTICAST_DEBUGGER_TRACE_HELPER, 2, 0);

        pCode->EmitBR(debuggerCheckEnd);
#endif
```

This should be even better

Comment (huoyaoyuan) [reply src/coreclr/vm/comdelegate.cpp:2209]: That was exactly what I did locally, and resulted in exact same codegen with current, at least for x86.

@dotnet/jit-contrib Do you have any suggestion on this?

File: src/coreclr/vm/comdelegate.cpp [hunk 2]
```diff
@@ -2166,7 +2159,11 @@ FCIMPL1(PCODE, COMDelegate::GetMulticastInvoke, Object* refThisIn)
 
         ILCodeStream *pCode = sl.NewCodeStream(ILStubLinker::kDispatch);
 
+#ifndef TARGET_X86
+        // x86 has much less available registers comparing to other architectures.
+        // To reduce register pressure and stack spilling, avoid saving invocationCount in local variable.
         DWORD dwInvocationCountNum = pCode->NewLocal(ELEMENT_TYPE_I4);
+#endif // TARGET_X86
```
Comment (jkotas) [inline src/coreclr/vm/comdelegate.cpp:2166]:
```suggestion
```
I think it is fine to do this for all platforms to keep it simple. Non-x86 platforms have to save an extra register in prolog/epilog for the local variable, so it boils down to whether to micro-optimize for small number of invocations or large number of invocations.

**Review by jkotas — COMMENTED** [review]
LGTM otherwise. Nice simplification!

File: src/coreclr/vm/comdelegate.cpp (hunk 2 above)
Comment (huoyaoyuan) [reply src/coreclr/vm/comdelegate.cpp:2166]: The difference looks negligible on x64. We should prefer simplification in this case.

File: src/coreclr/vm/comdelegate.cpp (hunk 1 above)
Comment (huoyaoyuan) [reply src/coreclr/vm/comdelegate.cpp:2209]: Well normalizing the loop closer to a `for` loop results in about 5% improvement for ManyArg case.

Comment (huoyaoyuan) [reply src/coreclr/vm/comdelegate.cpp:2200]: Codegen for x64 now:

```asm
; Assembly listing for method System.Action:IL_STUB_MulticastDelegate_Invoke():this (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; FullOpts code
; optimized code
; optimized using Synthesized PGO
; rsp based frame
; partially interruptible
; with Synthesized PGO: fgCalledCount is 100
; No PGO data

G_M000_IG01:                ;; offset=0x0000
       push     rsi
       push     rbx
       sub      rsp, 40
       mov      rbx, rcx

G_M000_IG02:                ;; offset=0x0009
       xor      esi, esi
       cmp      qword ptr [rbx+0x30], 0
       jle      SHORT G_M000_IG05

G_M000_IG03:                ;; offset=0x0012
       test     dword ptr [(reloc 0x7ff824c55210)], 512
       jne      SHORT G_M000_IG06

G_M000_IG04:                ;; offset=0x001E
       mov      rcx, gword ptr [rbx+0x28]
       cmp      esi, dword ptr [rcx+0x08]
       jae      SHORT G_M000_IG07
       mov      rax, gword ptr [rcx+8*rsi+0x10]
       mov      rcx, gword ptr [rax+0x08]
       call     [rax+0x18]System.Action:Invoke():this
       inc      esi
       movsxd   rcx, esi
       cmp      rcx, qword ptr [rbx+0x30]
       jl       SHORT G_M000_IG03

G_M000_IG05:                ;; offset=0x003E
       add      rsp, 40
       pop      rbx
       pop      rsi
       ret

G_M000_IG06:                ;; offset=0x0045
       mov      rcx, rbx
       mov      edx, esi
       call     System.StubHelpers.StubHelpers:MulticastDebuggerTraceHelper(System.Object,int)
       jmp      SHORT G_M000_IG04

G_M000_IG07:                ;; offset=0x0051
       call     CORINFO_HELP_RNGCHKFAIL
       int3

; Total bytes of code 87
```

File: src/coreclr/vm/comdelegate.cpp [hunk 3]
```diff
@@ -2166,38 +2159,41 @@ FCIMPL1(PCODE, COMDelegate::GetMulticastInvoke, Object* refThisIn)
...
-        pCode->EmitLoadThis();
-        pCode->EmitLDFLD(pCode->GetToken(CoreLibBinder::GetField(FIELD__MULTICAST_DELEGATE__INVOCATION_COUNT)));
-        pCode->EmitSTLOC(dwInvocationCountNum);
+        ILCodeLabel *checkCount = pCode->NewCodeLabel();
 
         // initialize counter
         pCode->EmitLDC(0);
         pCode->EmitSTLOC(dwLoopCounterNum);
 
+        // Make the shape of the loop similar to what C# compiler emits
```
Comment (jkotas) [inline src/coreclr/vm/comdelegate.cpp:2175]: The new shape seems to be missing the invocation of the trace helper at the end. E.g. if the invocation count is 2, the trace helper should be called 3 times. It is only called 2 times if I am reading the code correctly.

Comment (huoyaoyuan) [reply src/coreclr/vm/comdelegate.cpp:2175]: Fixed. Result codegen:

```asm
; Assembly listing for method System.Action:IL_STUB_MulticastDelegate_Invoke():this (FullOpts)
; Emitting BLENDED_CODE for X64 with AVX - Windows
; FullOpts code
; optimized code
; optimized using Synthesized PGO
; rsp based frame
; partially interruptible
; with Synthesized PGO: fgCalledCount is 100
; No PGO data

G_M000_IG01:                ;; offset=0x0000
       push     rsi
       push     rbx
       sub      rsp, 40
       mov      rbx, rcx

G_M000_IG02:                ;; offset=0x0009
       xor      esi, esi
       jmp      SHORT G_M000_IG04

G_M000_IG03:                ;; offset=0x000D
       mov      rcx, gword ptr [rbx+0x28]
       cmp      esi, dword ptr [rcx+0x08]
       jae      SHORT G_M000_IG08
       mov      rax, gword ptr [rcx+8*rsi+0x10]
       mov      rcx, gword ptr [rax+0x08]
       call     [rax+0x18]System.Action:Invoke():this
       inc      esi

G_M000_IG04:                ;; offset=0x0024
       test     dword ptr [(reloc 0x7ff84ea05210)], 512
       jne      SHORT G_M000_IG06

G_M000_IG05:                ;; offset=0x0030
       movsxd   rcx, esi
       cmp      rcx, qword ptr [rbx+0x30]
       jl       SHORT G_M000_IG03
       jmp      SHORT G_M000_IG07

G_M000_IG06:                ;; offset=0x003B
       mov      rcx, rbx
       mov      edx, esi
       call     System.StubHelpers.StubHelpers:MulticastDebuggerTraceHelper(System.Object,int)
       jmp      SHORT G_M000_IG05

G_M000_IG07:                ;; offset=0x0047
       add      rsp, 40
       pop      rbx
       pop      rsi
       ret

G_M000_IG08:                ;; offset=0x004E
       call     CORINFO_HELP_RNGCHKFAIL
       int3

; Total bytes of code 84
```

JIT reorders the debugging block in front of the `ret`. Benchmark shows no regression. It's just a bit strange with the unconditional jumps.

**Review by jkotas — APPROVED** [review]

**Pull request merged**